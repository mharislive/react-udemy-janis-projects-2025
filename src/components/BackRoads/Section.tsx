type SectionProps = React.HTMLAttributes<HTMLElement> & {
  children: React.ReactNode;
  title?: string;
  as?: keyof React.JSX.IntrinsicElements;
};

const Section = (props: SectionProps) => {
  const { children, title, className = "", as: Component = "section", id="" } = props;
  let formattedTitle: React.ReactNode = null;

  if (title) {
    const [first, ...rest] = title.trim().split(/\s+/);
    formattedTitle = (
      <h2 className="text-4xl font-bold text-center uppercase tracking-widest mb-16">
        {first}
        {rest.length > 0 && (
          <>
            {" "}
            <span className="text-cyan-500">{rest.join(" ")}</span>
          </>
        )}
      </h2>
    );
  }

  return (
    <Component className={`py-20 ${className}`} id={id}>
      {formattedTitle}
      {children}
    </Component>
  );
};

export default Section;
