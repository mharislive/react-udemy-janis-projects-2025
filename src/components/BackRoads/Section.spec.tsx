import { render, screen } from "@testing-library/react";
import { axe } from "vitest-axe";
import Section from "./Section";

describe("BackRoads - Section", () => {
  const SectionWithHeading = () => (
    <Section title="Heading" className="test" id="test">
      <p>Lorem Ipsum</p>
    </Section>
  );

  const SectionWithoutHeading = () => (
    <Section as="footer">
      <p>Lorem Ipsum</p>
    </Section>
  );

  it("should render successfully", () => {
    const { container } = render(<SectionWithHeading />);
    expect(container).toBeInTheDocument();
  });

  it("should render section tag by default", () => {
    const { container } = render(<SectionWithHeading />);
    expect(container.querySelector("section")).toBeInTheDocument();
  });

  it("should render footer tag if as props is set as footer", () => {
    const { container } = render(<SectionWithoutHeading />);
    expect(container.querySelector("footer")).toBeInTheDocument();
  });

  it("should render successfully with heading", () => {
    render(<SectionWithHeading />);
    expect(
      screen.getByRole("heading", { name: /heading/i })
    ).toBeInTheDocument();
  });

  it("should render successfully without heading", () => {
    render(<SectionWithoutHeading />);
    expect(screen.queryByRole("heading")).not.toBeInTheDocument();
  });

  it("should have the passed class to the parent container", () => {
    const { container } = render(<SectionWithHeading />);
    const section = container.querySelector("section");
    expect(section).toHaveClass("test");
  });

  it("should have the passed id to the parent container", () => {
    const { container } = render(<SectionWithHeading />);
    const section = container.querySelector("section");
    expect(section).toHaveAttribute("id", "test");
  });

  it("should render passed children successfully", () => {
    render(<SectionWithHeading />);
    expect(screen.getByText(/lorem ipsum/i)).toBeVisible();
  });

  it("should match snapshot with heading", () => {
    const { asFragment } = render(<SectionWithHeading />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should match snapshot without heading", () => {
    const { asFragment } = render(<SectionWithoutHeading />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should have no accessibility violations", async () => {
    let { container } = render(<SectionWithHeading />);
    let results = await axe(container);
    expect(results).toHaveNoViolations();

    ({ container } = render(<SectionWithoutHeading />));
    results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
