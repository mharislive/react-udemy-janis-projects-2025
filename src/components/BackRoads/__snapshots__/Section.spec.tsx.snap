// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BackRoads - Section > should match snapshot with heading 1`] = `
<DocumentFragment>
  <section
    class="py-20 test"
    id="test"
  >
    <h2
      class="text-4xl font-bold text-center uppercase tracking-widest mb-16"
    >
      Heading
    </h2>
    <p>
      Lorem Ipsum
    </p>
  </section>
</DocumentFragment>
`;

exports[`BackRoads - Section > should match snapshot without heading 1`] = `
<DocumentFragment>
  <footer
    class="py-20 "
    id=""
  >
    <p>
      Lorem Ipsum
    </p>
  </footer>
</DocumentFragment>
`;
